### PR TITLE
Fixes #35677 - Add ForemanGoogle plugin

### PR DESCRIPTION
**Pull request by stejskalleos**

### CODE REVIEW

**Comment by ekohl** [comment]
This will fail CI until a matching patch in https://github.com/theforeman/puppet-foreman is merged. Once the PR is available, please link it.

**Comment by stejskalleos** [comment]
Puppet PR: https://github.com/theforeman/puppet-foreman/pull/1040

**Comment by ekohl** [comment]
[test foreman-installer]

**Comment by evgeni** [comment]
[test foreman-installer]

File: config/foreman-answers.yaml
Comment (ekohl) [inline config/foreman-answers.yaml:26]: If you're dropping gce, you'll also need to reflect that in the migration. You could even use the existing answer to determine if the google plugin should be enabled.

**Review by ekohl — CHANGES_REQUESTED** [review]

**Comment by stejskalleos** [comment]
google rpm has been merged, rebased & updated

File: config/foreman.migrations/20220406150414_foreman_google.rb
Comment (ekohl) [inline config/foreman.migrations/20220406150414_foreman_google.rb]: Wouldn't it make sense to assume that if previously someone used gce, they now want the google plugin?

Comment (stejskalleos) [reply config/foreman.migrations/20220406150414_foreman_google.rb]: Yop, make sense, updated.

However `rspec ./spec/migration_spec.rb` tests are still failing on mismatch with answers, not sure why

File: config/katello.migrations/20220406141823_foreman_google.rb
Comment (ehelms) [inline config/katello.migrations/20220406141823_foreman_google.rb]: Need the same logic here as the foreman migration

File: config/foreman.migrations/20220406150414_foreman_google.rb
Comment (ekohl) [inline config/foreman.migrations/20220406150414_foreman_google.rb]: `Hash#delete` returns the value so you can also write this as:
```suggestion
answers['foreman::plugin::google'] = answers.delete('foreman::compute::gce')
```

File: config/katello.migrations/20220406141823_foreman_google.rb
Comment (ekohl) [inline config/katello.migrations/20220406141823_foreman_google.rb]:
```suggestion
answers['foreman::plugin::google'] = answers.delete('foreman::compute::gce')
```

**Review by ekohl — CHANGES_REQUESTED** [review]

**Comment by stejskalleos** [comment]
@ekohl PR updated, however `rspec ./spec/migration_spec.rb` tests are still failing on mismatch with answers, not sure why

File: config/foreman.migrations/20220406150414_foreman_google.rb
Comment (ekohl) [inline config/foreman.migrations/20220406150414_foreman_google.rb]: Looking closer at it, the problem is that this migration runs on `config/foreman-answers.yaml`. There is no answer for `foreman::compute::gce` so it's nil. That means it is essentially:
```ruby
answers['foreman::plugin::google'] = nil
```

I think a safe version would be:
```suggestion
old_answer = answers.delete('foreman::compute::gce')
answers['foreman::plugin::google'] ||= old_answer || false
```

The latter part (`|| false`) because `false || nil` is `nil`. The former part `||=` to respect if the answer is already `true`. The delete is on its own line because it must happen and if the answer is already true, it won't be evaluated.

**Comment by stejskalleos** [comment]
@ekohl migration files updated but it still fails :/

File: spec/fixtures/pulpcore-migration/katello-answers-after.yaml
Comment (ekohl) [inline spec/fixtures/pulpcore-migration/katello-answers-after.yaml:24]: This is a modification that's missing in the other fixtures, causing the tests to fail.

**Comment by stejskalleos** [comment]
@ekohl and it's :green_apple: !

**Review by ekohl — CHANGES_REQUESTED** [review]
Can you include https://github.com/theforeman/foreman-installer/pull/815 here? Then also do some additional logic in the migration that if `foreman::cli` and `foreman::plugin::google` it also enabled `foreman::cli::google` so users have a smooth migration?

**Comment by stejskalleos** [comment]
@ekohl https://github.com/theforeman/foreman-installer/pull/815 added

**Comment by stejskalleos** [comment]
[test foreman-installer]

**Comment by stejskalleos** [comment]
Tests are failing on `[Librarian] Conflict between theforeman-pulpcore (>= 7.0.1, < 8.0.0) <https://forgeapi.puppet.com/> (from katello-foreman_proxy_content) and theforeman-pulpcore/7.0.0 <https://github.com/theforeman/puppet-pulpcore#master>
Could not resolve the dependencies.`, not sure if it is related to my changes

File: config/foreman.migrations/20220406150414_foreman_google.rb
Comment (ekohl) [inline config/foreman.migrations/20220406150414_foreman_google.rb]: Users can have disabled the CLI and if you then enable a CLI plugin it'll fail. I think this should take care of it.
```suggestion
answers['foreman::cli::google'] ||= answers['foreman::cli'] && old_answer || false
```

But what I'd prefer is to be more explicit in the logic, which is to enable it if the CLI and Google plugin are installed:
```suggestion
answers['foreman::cli::google'] ||= answers['foreman::cli'] && answers['foreman::plugin::google']
```

And copy the same to the Katello migration.

File: spec/fixtures/pulpcore-migration-dont-use-content-plugins-on-upgrades/katello-answers-before.yaml
Comment (ekohl) [inline spec/fixtures/pulpcore-migration-dont-use-content-plugins-on-upgrades/katello-answers-before.yaml]: This can remain. It was a snapshot of how the migration was back then.

File: spec/fixtures/pulpcore-migration-rpm-only/katello-answers-before.yaml
Comment (ekohl) [inline spec/fixtures/pulpcore-migration-rpm-only/katello-answers-before.yaml]: And this can remain too.

**Comment by ekohl** [comment]
No, that pulpcore failure was because I should have done https://github.com/theforeman/puppet-pulpcore/commit/e124f89c117d221ecd350daf2d0627df4a573531 before merging the other PR.

**Comment by stejskalleos** [comment]
@ekohl cool, now it's green again

**Review by ekohl — APPROVED** [review]
I think this is correct now. I'd say it's now blocked on packaging.

**Pull request merged**
